### PR TITLE
Add memory_format in torch.Tensor.to

### LINKED ISSUE #144: Support `memory_format` on `to()`
## 🚀 Feature

`to(memory_format=something)` is part of the `MegatronImagen` model in NeMo.

Ideally, this would work:
```
$ git diff .
diff --git a/nemo/collections/multimodal/models/text_to_image/imagen/imagen.py b/nemo/collections/multimodal/models/text_to_image/imen/imagen.py
index 4fa6cd230..2cf7a8ffa 100644
--- a/nemo/collections/multimodal/models/text_to_image/imagen/imagen.py
+++ b/nemo/collections/multimodal/models/text_to_image/imagen/imagen.py
@@ -31,6 +31,7 @@ from nemo.collections.nlp.modules.common.megatron.module import Float16Module
 from nemo.collections.nlp.parts.utils_funcs import get_last_rank
 from nemo.core.classes.common import Serialization
 from nemo.utils import logging
+import thunder
 
 try:
     from apex import amp
@@ -190,6 +191,7 @@ class MegatronImagen(MegatronBaseModel):
         self.megatron_amp_O2 = cfg.get('megatron_amp_O2', False)
 
         self.model = self.model_provider_func()
+        self.model = thunder.jit(self.model)
 
         if self.trainer.precision in ['bf16', 'bf16-mixed']:
             self.autocast_dtype = torch.bfloat16
```

### Motivation

Trying to evaluate NeMo models in thunder and expand our model support there. Megatron-based models appear to be widely used.

### Alternatives

I wonder if we could temporarily just accept the keyword without actually doing anything about it. I imagine that would be very slow, but it might allow us to get models like this one into thunder more easily.

I'll start trying to convert smaller parts of the model next.

### Additional context

Model in question:

https://github.com/NVIDIA/NeMo/blob/23baa48e441ecb6cc6b49c23bf8cfc076db38bdc/nemo/collections/multimodal/models/text_to_image/imagen/imagen.py#L175

I think the `to` that is failing for me 
is actually this line:
https://github.com/NVIDIA/NeMo/blob/23baa48e441ecb6cc6b49c23bf8cfc076db38bdc/nemo/collections/multimodal/models/text_to_image/imagen/imagen.py#L135

Model test:
[log.txt](https://github.com/Lightning-AI/lightning-thunder/files/14912108/log.txt)

### PR DESCRIPTION
**Pull request by jjsjann123**
## What does this PR do?

Fixes #144.

Adding argument `memory_format` in `torch.Tensor.to`.

### CODE REVIEW

**Comment by tfogal** [comment]
Hey @jjsjann123, is it possible this is coming from somewhere else?

with:

> commit eb42bdb7624385d53237d0737a10b334ee2e79d8 (HEAD, jie/to_memory_format)
> Author: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
> Date:   Wed Apr 10 23:08:55 2024 +0000
> 
>     [pre-commit.ci] auto fixes from pre-commit.com hooks
>     
>     for more information, see https://pre-commit.ci

 I am still getting:
```
[rank0]:     super().optimizer_step(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/core/module.py", line 1303, in optimizer_step
[rank0]:     optimizer.step(closure=optimizer_closure)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/core/optimizer.py", line 152, in step
[rank0]:     step_output = self._strategy.optimizer_step(self._optimizer, closure, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/strategies/ddp.py", line 270, in optimizer_step
[rank0]:     optimizer_output = super().optimizer_step(optimizer, closure, model, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/strategies/strategy.py", line 239, in optimizer_step
[rank0]:     return self.precision_plugin.optimizer_step(optimizer, model=model, closure=closure, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/plugins/precision/amp.py", line 80, in optimizer_step
[rank0]:     closure_result = closure()
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/loops/optimization/automatic.py", line 144, in __call
__
[rank0]:     self._result = self.closure(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
[rank0]:     return func(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/loops/optimization/automatic.py", line 129, in closur
e
[rank0]:     step_output = self._step_fn()
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/loops/optimization/automatic.py", line 318, in _train
ing_step     
[rank0]:     training_step_output = call._call_strategy_hook(trainer, "training_step", *kwargs.values())
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/call.py", line 309, in _call_strategy_hook
[rank0]:     output = fn(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/strategies/strategy.py", line 390, in training_step
[rank0]:     return self._forward_redirection(self.model, self.lightning_module, "training_step", *args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/strategies/strategy.py", line 642, in __call__
[rank0]:     wrapper_output = wrapper_module(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1541, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/strategies/strategy.py", line 635, in wrapped_forward
[rank0]:     out = method(*_args, **_kwargs)
[rank0]:   File "/home/tfogal/dev/nemo/nemo/utils/model_utils.py", line 399, in wrap_training_step
[rank0]:     output_dict = wrapped(*args, **kwargs)
[rank0]:   File "/home/tfogal/dev/nemo/nemo/collections/multimodal/models/text_to_image/imagen/imagen.py", line 393, in training
_step
[rank0]:     loss_mean, loss_dict = self.fwd_bwd_step(dataloader_iter, False)
[rank0]:   File "/home/tfogal/dev/nemo/nemo/collections/multimodal/models/text_to_image/imagen/imagen.py", line 346, in fwd_bwd_
step
[rank0]:     losses_reduced_per_micro_batch = fwd_bwd_function(
[rank0]:   File "/home/tfogal/dev/mlm/megatron/core/pipeline_parallel/schedules.py", line 374, in forward_backward_no_pipelining
[rank0]:     output_tensor = forward_step(
[rank0]:   File "/home/tfogal/dev/mlm/megatron/core/pipeline_parallel/schedules.py", line 192, in forward_step
[rank0]:     output_tensor, loss_func = forward_step_func(data_iterator, model)
[rank0]:   File "/home/tfogal/dev/nemo/nemo/collections/multimodal/models/text_to_image/imagen/imagen.py", line 254, in fwd_outp
ut_and_loss_func
[rank0]:     loss, loss_dict = model(*batch)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1541, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/__init__.py", line 194, in forward
[rank0]:     res = self._forward_fn(*args, **kwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/__init__.py", line 629, in fn_
[rank0]:     cache_entry, inps, pro_to_epi = get_computation_and_inputs(*args, **kwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/__init__.py", line 262, in cache_info_wrapper
[rank0]:     res = fn(*args, **kwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/__init__.py", line 504, in get_computation_and_inputs
[rank0]:     prologue_trc, computation_trc, *maybe_epilogue = interpreter(
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/__init__.py", line 175, in _general_frontend
[rank0]:     return thunder_general_jit(fn, args, kwargs, sharp_edges=sharp_edges)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/jit_ext.py", line 1440, in thunder_general_jit
[rank0]:     result = jfn(*args, **kwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/interpreter.py", line 6684, in fn_
[rank0]:     raise e
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/interpreter.py", line 6647, in fn_2
[rank0]:     return fn(*args, **kwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/interpreter.py", line 6046, in _impl
[rank0]:     return fn.__func__(fn.__self__, *args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/interpreter.py", line 6046, in _impl
[rank0]:     return fn.__func__(fn.__self__, *args, **kwargs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1541, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/interpreter.py", line 6046, in _impl
[rank0]:     return fn.__func__(fn.__self__, *args, **kwargs)
[rank0]:   File "/home/tfogal/dev/nemo/nemo/collections/multimodal/models/text_to_image/imagen/imagen.py", line 135, in forward
[rank0]:     x_start = x_start.to(memory_format=torch.channels_last)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/interpreter.py", line 6158, in partial_call_impl
[rank0]:     return partial_function.func(*(partial_function.args + args), **(partial_function.keywords | kwargs))
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/interpreter.py", line 1258, in wrapping_wrapper
[rank0]:     res = ufn(*uargs, **ukwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/symbol.py", line 250, in __call__
[rank0]:     result = self.meta(*args, **kwargs)
[rank0]:   File "/opt/pytorch/lightning-thunder/thunder/core/langctxs.py", line 124, in _fn
[rank0]:     result = fn(*args, **kwargs)
[rank0]: TypeError: to() got an unexpected keyword argument 'memory_format'
```

from the NeMo model. Did I apply it wrong somehow?

**Comment by jjsjann123** [comment]
Interesting. sounds like my patch isn't bullet proof then.
Is it easy to run the repro? I can take a look there.

**Comment by tfogal** [comment]
> Is it easy to run the repro? I can take a look there.

"Easy" might be a bit of a strong term... i'll ping you on slack, you'd need something internal.

File: thunder/torch/__init__.py
Comment (crcrpar) [inline thunder/torch/__init__.py:1]: it'd be better to update [TODO of L239](https://github.com/Lightning-AI/lightning-thunder/pull/157/files#diff-0f7bc0b6dc143555fbe3e597d9854ba7b99a2a33f454deeea9766e2d8e2da697L239)

Comment (jjsjann123) [reply thunder/torch/__init__.py:1]: Thanks for double checking this one :bow:

**Comment by jjsjann123** [comment]
```
FAILED thunder/tests/test_inplace_copy.py::test_prim_inplace_copy_fwd_nvfuser_cuda_bfloat16 - AssertionError: Tensor-likes are not close!

Mismatched elements: 1 / 16 (6.2%)
Greatest absolute difference: 0.01953125 at index (1, 1) (up to 0.01 allowed)
Greatest relative difference: 0.032958984375 at index (1, 1) (up to 0.01 allowed)
```

Is this a numerical issue? since I'm not reproing this one locally.

**Comment by kiya00** [comment]
> ```
> FAILED thunder/tests/test_inplace_copy.py::test_prim_inplace_copy_fwd_nvfuser_cuda_bfloat16 - AssertionError: Tensor-likes are not close!
> 
> Mismatched elements: 1 / 16 (6.2%)
> Greatest absolute difference: 0.01953125 at index (1, 1) (up to 0.01 allowed)
> Greatest relative difference: 0.032958984375 at index (1, 1) (up to 0.01 allowed)
> ```
> 
> Is this a numerical issue? since I'm not reproing this one locally.

I think so, I've already relaxed this bf16 to 1e-2, I'll check what's wrong, you could re-run this test I think

**Comment by jjsjann123** [comment]
> > ```
> > FAILED thunder/tests/test_inplace_copy.py::test_prim_inplace_copy_fwd_nvfuser_cuda_bfloat16 - AssertionError: Tensor-likes are not close!
> > 
> > Mismatched elements: 1 / 16 (6.2%)
> > Greatest absolute difference: 0.01953125 at index (1, 1) (up to 0.01 allowed)
> > Greatest relative difference: 0.032958984375 at index (1, 1) (up to 0.01 allowed)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Is this a numerical issue? since I'm not reproing this one locally.
> 
> I think so, I've already relaxed this bf16 to 1e-2, I'll check what's wrong, you could re-run this test I think

Thanks for confirming.
I still have Tom's use case to clear but struggling with running that, so not really blocked by the CI yet.

**Comment by tfogal** [comment]
> Is it easy to run the repro? I can take a look there.

Jie and I iterated offline.

This was my bad. I was running things incorrectly (had 2 thunder installs, picked up wrong one). The patch works fine.

Sorry for the noise!

File: thunder/torch/__init__.py [hunk 1]
```diff
@@ -259,6 +260,12 @@ def to(
         if dtype is not None:
             dtype = to_dtype(dtype)
             a = prims.convert_element_type(a, dtype)
+        if memory_format is not None:
+            # NOTE not sure if we need to handle torch.preserve_format explicitly
```
Comment (mruberry) [inline thunder/torch/__init__.py:264]: What does copying to a different device (like CUDA -> CPU) do when the original tensor is noncontiguous?

Comment (jjsjann123) [reply thunder/torch/__init__.py:264]: > when the original tensor is noncontiguous?

I think the protocol here is that, if a copy is happening, it's always going to be packed tensor at the destination.
The original tensor only suggests a layout and whether or not it packed isn't important.

Comment (jjsjann123) [reply thunder/torch/__init__.py:264]: > What does copying to a different device (like CUDA -> CPU) 

The behavior should also be the same whether it's moving across device or on the same device with `copy == True`.

Comment (jjsjann123) [reply thunder/torch/__init__.py:264]:
```
>>> y = x.to(copy=True, memory_format=torch.preserve_format)
>>> y.stride()
(36, 12, 3, 1)
>>> x.stride()
(60, 20, 5, 2)
```

**Review by t-vi — APPROVED** [review]
Thank you @jjsjann123 @tfogal

**Pull request merged**